transports/janus_pfunix.c

```diff
@@ -0,0 +1,526 @@
+			poll_fds[fds].fd = pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		if(admin_pfd > -1) {
+			/* Admin API */
+			poll_fds[fds].fd = admin_pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		/* Iterate on available clients, to see if we need to POLLIN or POLLOUT too */
+		janus_mutex_lock(&clients_mutex);
+		GHashTableIter iter;
+		gpointer value;
+		g_hash_table_iter_init(&iter, clients_by_fd);
+		while(g_hash_table_iter_next(&iter, NULL, &value)) {
+			janus_pfunix_client *client = value;
+			if(client->fd > -1) {
+				poll_fds[fds].fd = client->fd;
+				poll_fds[fds].events = g_async_queue_length(client->messages) > 0 ? POLLIN | POLLOUT : POLLIN;
+				fds++;
+			}
+		}
+		janus_mutex_unlock(&clients_mutex);
+
+		/* Start polling */
+		int res = poll(poll_fds, fds, -1);
+		if(res == 0)
+			continue;
+		if(res < 0) {
+			JANUS_LOG(LOG_ERR, "poll() failed\n");
+			break;
+		}
+		for(int i=0; i<fds; i++) {
+			if((poll_fds[i].revents & POLLOUT) == POLLOUT) {
+				/* Find the client from its file descriptor */
+				janus_mutex_lock(&clients_mutex);
+				janus_pfunix_client *client = g_hash_table_lookup(clients_by_fd, GINT_TO_POINTER(poll_fds[i].fd));
+				if(client != NULL) {
+					char *payload = NULL;
+					while((payload = g_async_queue_try_pop(client->messages)) != NULL) {
+						int res = 0;
+						do {
+							res = write(client->fd, payload, strlen(payload));
+						} while(res == -1 && errno == EINTR);
+						/* FIXME Should we check if sent everything? */
+						JANUS_LOG(LOG_HUGE, "Written %d/%zu bytes on %d\n", res, strlen(payload), client->fd);
+						g_free(payload);
+					}
+				}
+				janus_mutex_unlock(&clients_mutex);
+			}
+			if((poll_fds[i].revents & POLLIN) == POLLIN) {
+				if(poll_fds[i].fd == write_fd[0]) {
+					/* Read and ignore: we use this to unlock the poll if there's data to write */
+					res = read(poll_fds[i].fd, buffer, BUFFER_SIZE);
+				} else if(poll_fds[i].fd == pfd || poll_fds[i].fd == admin_pfd) {
+					/* Janus/Admin API: accept the new client */
+					struct sockaddr_un address;
+					socklen_t address_length = 0;
+					int cfd = accept(pfd, (struct sockaddr *) &address, &address_length);
+					if(cfd > -1) {
+						JANUS_LOG(LOG_INFO, "Got new Unix Sockets %s API client: %d\n",
+							poll_fds[i].fd == pfd ? "Janus" : "Admin", cfd);
+						/* Allocate new client */
+						janus_pfunix_client *client = g_malloc0(sizeof(janus_pfunix_client));
+						client->fd = cfd;
+						client->admin = (poll_fds[i].fd == admin_pfd);	/* API client type */
+						client->messages = g_async_queue_new();
+						client->session_timeout = 0;
+						/* Take note of this new client */
+						janus_mutex_lock(&clients_mutex);
+						g_hash_table_insert(clients_by_fd, GINT_TO_POINTER(cfd), client);
+						g_hash_table_insert(clients, client, client);
+						janus_mutex_unlock(&clients_mutex);
+					}
+				} else {
+					/* Client data: receive */
+					res = read(poll_fds[i].fd, buffer, BUFFER_SIZE);
```

Comment:
I wasn't aware of this possibility, thanks! Looking at the docs, you'd get a MSG_TRUNC back, which means we might then try again, later or right away. The only issue with this is that, being single threaded, I'll have to handle the fragmentation of the message in a subsequent poll myself, which means keeping info on partial messages in the client structure (pretty much as we do in the websockets transport). I was hoping I could get away with this "for free" with Unix Sockets but better safe than sorry :-) What I haven't clear is, can I get the rest through another ```recvmsg()```, or do I need something else? Would I get the whole message by passing a larger buffer, or only what was left to read?